File: Dynamic-Programming/1035-M-Uncrossed-Lines/uncrossed_lines.cpp
class Solution {
public:
    int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
        int m = nums2.size();
        int n = nums1.size();

        vector<vector<int>> OPT(m+1, vector<int>(n+1, 0));

        for (int i=1; i<m+1; i++) {
            for (int j=1; j<n+1; j++) {
                if (nums1[j-1] == nums2[i-1]) {
                    OPT[i][j] = 1 + OPT[i-1][j-1];
                } else {
                    OPT[i][j] = max(OPT[i-1][j], OPT[i][j-1]);
                }
            }
        }
        return OPT[m][n];
    }
};

File: Dynamic-Programming/1035-M-Uncrossed-Lines/uncrossed_lines.py
class Solution:
    def maxUncrossedLines(self, nums1: list[int], nums2: list[int]) -> int:
        n = len(nums1)
        m = len(nums2)

        OPT = [[0]*(n+1) for _ in range(m+1)]

        for i in range(1, m+1):
            for j in range(1, n+1):
                if nums2[i-1] == nums1[j-1]:
                    OPT[i][j] = 1 + OPT[i-1][j-1]
                else:
                    OPT[i][j] = max(OPT[i-1][j], OPT[i][j-1])
        return OPT[m][n]